src/mapstorage/XmlMapStorage.cpp

```diff
@@ -0,0 +1,462 @@
+    stream.writeEndElement(); // end exit
+}
+
+void XmlMapStorage::saveExitTo(QXmlStreamWriter &stream, const Exit &e)
+{
+    QString idlist;
+    const char *separator = nullptr;
+    for (const RoomId id : e.outRange()) {
+        if (separator == nullptr) {
+            separator = ",";
+        } else {
+            idlist += separator;
+        }
+        idlist += QString("%1").arg(id.asUint32());
+    }
+    stream.writeAttribute("to", idlist);
+}
+
+void XmlMapStorage::saveMarkers(QXmlStreamWriter &stream, const MarkerList &markerList)
+{
+    ProgressCounter &progressCounter = getProgressCounter();
+    for (const auto &marker : markerList) {
+        saveMarker(stream, deref(marker));
+        progressCounter.step();
+    }
+}
+
+void XmlMapStorage::saveMarker(QXmlStreamWriter &stream, const InfoMark &marker)
+{
+    const InfoMarkTypeEnum type = marker.getType();
+    stream.writeStartElement("marker");
+    stream.writeAttribute("type", markTypeName(type));
+    stream.writeAttribute("class", markClassName(marker.getClass()));
+    // REVISIT: round to 45 degrees?
+    if (marker.getRotationAngle() != 0) {
+        stream.writeAttribute("angle",
+                              QString("%1").arg(static_cast<qint32>(marker.getRotationAngle())));
+    }
+    saveCoordinate(stream, "pos1", marker.getPosition1());
+    saveCoordinate(stream, "pos2", marker.getPosition2());
+
+    if (type == InfoMarkTypeEnum::TEXT) {
+        stream.writeStartElement("text");
+        stream.writeCharacters(marker.getText().toQString());
+        stream.writeEndElement(); // end text
+    }
+
+    stream.writeEndElement(); // end marker
+}
+
+void XmlMapStorage::saveXmlElement(QXmlStreamWriter &stream,
+                                   const QString &name,
+                                   const QString &value)
+{
+    if (!value.isEmpty()) {
+        stream.writeStartElement(name);
+        stream.writeCharacters(value);
+        stream.writeEndElement(); // end description
+    }
+}
+
+void XmlMapStorage::saveXmlAttribute(QXmlStreamWriter &stream,
+                                     const QString &name,
+                                     const QString &value)
+{
+    if (!value.isEmpty()) {
+        stream.writeAttribute(name, value);
+    }
+}
+
+void XmlMapStorage::log(const QString &msg)
+{
+    emit sig_log("XmlMapStorage", msg);
+}
+
+void XmlMapStorage::saveDoorFlags(QXmlStreamWriter &stream, DoorFlags fl)
+{
+    if (fl.isEmpty()) {
+        return;
+    }
+    for (uint x = 0; x < NUM_DOOR_FLAGS; x++) {
+        const DoorFlagEnum e = DoorFlagEnum(x);
+        if (fl.contains(e)) {
+            saveXmlElement(stream, "doorflag", doorFlagName(e));
+        }
+    }
+}
+
+void XmlMapStorage::saveExitFlags(QXmlStreamWriter &stream, ExitFlags fl)
+{
+    fl.remove(ExitFlagEnum::EXIT); // always set, do not save it
+    if (fl.isEmpty()) {
+        return;
+    }
+    for (uint x = 0; x < NUM_EXIT_FLAGS; x++) {
```

Comment:
You can use X_FOREACH_EXIT_FLAG instead

Reply:
Another alternative are the enums in https://github.com/MUME/MMapper/blob/master/src/mapdata/enums.h

Reply:
I am not sure to understand how you propose to write the loop on all `ExitFlagEnum` values.
Maybe you mean something like
```C++
#define X(UPPER, ...) ExitFlagEnum::UPPER,
for (ExitFlagEnum e : { X_FOREACH_EXIT_FLAG(x) }) {
    // ...
}
#undef X
```
using an implicit `std::initializer_list<ExitFlagEnum>` ?
Possible, but the heavy use of macros in C++ is really against my usual style...

Reply:
I'll check how to use the macros `ALL_EXIT_FLAGS` and friends, as per your suggestion. If they allow writing
```
for (ExitFlagEnum e : ALL_EXIT_FLAGS) {
    // ...
}
```
then it's definitely cleaner :)

Reply:
See if the enums work better. They haven't been used very much in the codebase yet but they exist for this purpose :-)

Reply:
Done for `DoorFlagEnum`, `ExitFlagEnum`, `RoomLoadFlagEnum` and `RoomMobFlagEnum`.
The new iteration
```C++
for (const ExitFlagEnum e : ALL_EXIT_FLAGS) {
 // ...
}
```
works and is indeed cleaner